### PR TITLE
feat: Remove `admin` permissions for ArgoCD instance when namespace label is removed

### PR DESCRIPTION
**Pull request by shubhamagarwal19**
**What type of PR is this?**
/kind enhancement

**What does this PR do / why we need it**:
This PR provides the ability to remove the permissions to manage a namespace from an argocd instance when the Label `argocd.argoproj.io/managed-by` is removed. 

**Have you updated the necessary documentation?**

* [ ] Documentation update is required by this PR.
* [ ] Documentation has been updated.

**Which issue(s) this PR fixes**:

Fixes #?

**How to test changes / Special notes to the reviewer**:
- Deploy the latest Operator.
- Create an ArgoCD instance.
- Create a new namespace with the label `argocd.argoproj.io/managed-by: <argocd-instance-namespace>`
- Make sure the ArgoCD instance has the permissions to manage the namespace by deploying an application or verifying if the correct roles/role_bindings were created.
- Manually remove the label `argocd.argoproj.io/managed-by: <argocd-instance-namespace>` from the namespace.
- Make sure the created roles/rolebindings were deleted. 
- Commit some change to the github repo and trigger the sync. It should fail.

### CODE REVIEW

File: pkg/controller/argocd/util.go [hunk 1]
```diff
@@ -1067,7 +1070,82 @@ func namespaceFilterPredicate() predicate.Predicate {
...
 	}
 }
+
+func deleteRBACsForNamespace(ownerNS, sourceNS string) error {
+	log.Info(fmt.Sprintf("Removing the RBACs created for the namespace: %s", sourceNS))
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		log.Error(err, fmt.Sprintf("unable to get k8s config"))
+		return err
+	}
```
Comment (jannfis) [inline pkg/controller/argocd/util.go:1090]: Will this work when the Operator is run locally? We had that issue before I think.

**Review by jannfis — COMMENTED** [review]
This code is on a very critical path and has _nil_ explanation of what it's doing. Please put out in comments what it is doing, and why it is doing that.

File: pkg/controller/argocd/util.go (hunk 1 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go]: Thanks Jann!! Completely forgot about that. Fixed it!! Also added comments

File: pkg/controller/argocd/util.go [hunk 2]
```diff
@@ -1064,10 +1067,98 @@ func containsString(arr []string, s string) bool {
...
+	for _, roleBinding := range roleBindings.Items {
+		err = k8sClient.RbacV1().RoleBindings(sourceNS).Delete(context.TODO(), roleBinding.Name, metav1.DeleteOptions{})
+		if err != nil {
+			log.Error(err, fmt.Sprintf("failed to delete role binding for namespace: %s", sourceNS))
+		}
+	}
+
+	// Get the cluster secret used for configuring ArgoCD
+	labelSelector = metav1.LabelSelector{MatchLabels: map[string]string{common.ArgoCDSecretTypeLabel: "cluster"}}
+	secrets, err := k8sClient.CoreV1().Secrets(ownerNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
```
Comment (jannfis) [inline pkg/controller/argocd/util.go:1137]: I think it's better to get the secret by name. User might have configured remote clusters (which end up having a `cluster` secret-type label as well). Now think of the situation where a remote cluster has a similar namespace to the one we are going to remove from the in-cluster secret -> oops.

File: pkg/controller/argocd/util.go [hunk 3]
```diff
@@ -1064,10 +1067,98 @@ func containsString(arr []string, s string) bool {
...
+	// List all the roles created for ArgoCD using the label selector
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{common.ArgoCDKeyPartOf: common.ArgoCDAppName}}
+	roles, err := k8sClient.RbacV1().Roles(sourceNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to list roles for namespace: %s", sourceNS))
+		return err
+	}
+
+	// Delete all the retrieved roles
+	for _, role := range roles.Items {
```
Comment (jannfis) [inline pkg/controller/argocd/util.go:1113]: We do set `ownerReferences` to the roles and rolebindings we create, right? I think it would make sense to check whether we are the owner, before deleting the role. I can imagine a scenario where an operator or even a user adds the part-of label with a value of `argocd` manually, and then this role or rolebinding would get removed by accident.

File: pkg/controller/argocd/util.go [hunk 4]
```diff
@@ -1064,10 +1067,98 @@ func containsString(arr []string, s string) bool {
...
+
+	// List all the roles bindings created for ArgoCD using the label selector
+	roleBindings, err := k8sClient.RbacV1().RoleBindings(sourceNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to list role bindings for namespace: %s", sourceNS))
+		return err
+	}
+
+	// Delete all the retrieved role bindings
+	for _, roleBinding := range roleBindings.Items {
```
Comment (jannfis) [inline pkg/controller/argocd/util.go:1128]: Same as above for roles.

**Review by jannfis — CHANGES_REQUESTED** [review]
Thank you @shubhamagarwal19. LGTM so far, only have a few concerns. Please see below.

File: pkg/controller/argocd/util.go [hunk 5]
```diff
@@ -1064,10 +1067,98 @@ func containsString(arr []string, s string) bool {
...
+		return err
+	}
+	for _, secret := range secrets.Items {
+		if namespaces, ok := secret.Data["namespaces"]; ok {
+			namespaceList := strings.Split(string(namespaces), ",")
+			var result []string
+
+			for _, n := range namespaceList {
+				// remove the namespace from the list of namespaces
+				if strings.TrimSpace(n) == sourceNS {
```
Comment (wtam2018) [inline pkg/controller/argocd/util.go:1149]: If this is the only namespace on the list and we remove it, we end up with an empty namespaces slice which means all namespaces.  Will the (non-default) argocd instance in this namespace have problem to create/sync applications in the same namespace (because it tries to access all namespaces)?

File: pkg/controller/argocd/util.go [hunk 6]
```diff
@@ -1064,10 +1067,98 @@ func containsString(arr []string, s string) bool {
...
+			}
+			// Update the secret with the updated list of namespaces
+			if _, err = k8sClient.CoreV1().Secrets(ownerNS).Update(context.TODO(), &secret, metav1.UpdateOptions{}); err != nil {
+				log.Error(err, fmt.Sprintf("failed to update cluster permission secret for namespace: %s", ownerNS))
+				return err
+			}
+		}
+	}
+	return nil
+}
```
Comment (wtam2018) [inline pkg/controller/argocd/util.go:1164]: How about adding some unit tests?

File: pkg/controller/argocd/util.go (hunk 2 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go:1137]: Maybe we can do the cluster address check ? if it is `https://kubernetes.default.svc` or not ? 
We can't do a name check because there can be a case where user has already created the secret with the configuration using the manual workaround we provided earlier. So we can't be sure of the name ? 
Also, we are doing a similar check while creating the secret as well, so i think cluster address check should be fine?

File: pkg/controller/argocd/util.go (hunk 5 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go:1149]: The implementation ensures that the secret always contains the namespace where argocd instance is deployed, i.e if ArgoCD was installed in `foo` namespace, the secret would always have `foo` namespace in the list of namespaces until the ArgoCD instance is marked as a `CLUSTER_CONFIG_INSTANCE`, then it would remove the list of namespaces from the secret.

Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go:1149]: so even if a user tries to remove the label or update the label to a different value, reconciler will make sure that the label is added again.

File: pkg/controller/argocd/util.go (hunk 3 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go:1113]: Yep, even i had thought of using the `managed-by` label with the value as `argocd-instance-name`, but we don't have the ArgoCD instance details. So i went with `part-of` label instead. 
Let me check if i can somehow retireve the ArgoCD instance name!

File: pkg/controller/argocd/util.go (hunk 6 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go:1164]: Yeah, i tried writing the UTs for it, but it is calling multiple functions like `config.GetConfig()` and `kubernetes.NewForConfig()`, which i am not sure how to mock. 
I also checked other pieces of code which is calling the same function, that doesn't have any UTs either which i could reference!
So, still exploring on this!

File: pkg/controller/argocd/util.go (hunk 3 above)
Comment (jannfis) [reply pkg/controller/argocd/util.go:1113]: I think it'd be sufficient for now to check for the `ownerReferences` being an `ArgoCD` resource.

File: pkg/controller/argocd/util.go (hunk 2 above)
Comment (jannfis) [reply pkg/controller/argocd/util.go:1137]: Yes, checking for secret being for `https://kubernetes.default.svc` (or rather, the appropriate constant) would be sufficient.

Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go:1137]: done!

File: pkg/controller/argocd/util.go (hunk 3 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go:1113]: As discussed! we can't have a cross-namespace owner reference when the owner itself is namespace scoped. 
So, the resources created by the operator don't have the owner references which could be evaluated. 

Will add a story to track this for the future!

Comment (jannfis) [reply pkg/controller/argocd/util.go:1113]: Thanks, and you're correct with your finding about cross-namespace ownerReferences. The risk is for edge cases only, so a future implementation is fine for me.

**Review by jannfis — APPROVED** [review]
Thanks, LGTM now.

Please add unit tests as @wtam2018 suggested before merging.

File: pkg/controller/argocd/util.go (hunk 6 above)
Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go:1160]: Added the unit test cases !!

File: pkg/controller/argocd/util.go [hunk 7]
```diff
@@ -1064,10 +1067,110 @@ func containsString(arr []string, s string) bool {
...
+
+func initK8sClient() (*kubernetes.Clientset, error) {
```
Comment (jannfis) [inline pkg/controller/argocd/util.go:1162]: This seems like a dupe of https://github.com/argoproj-labs/argocd-operator/blob/98a64dadb9b1687302d8de81cfe68fb06af2404e/pkg/reconciler/openshift/openshift.go#L346

It makes sense to move above method to a dedicated package and make it public.
